### DISCUSSION TITLE
How to recreate the model using the best suggested model settings?

### DISCUSSION THREAD

**Discussion by vinay-k12** [microsoft/FLAML#889] [General] [answered]
I ran the estimator and found the following settings as most suitable for my data: 

Best hyperparmeter config: {'n_estimators': 138, 'max_leaves': 37, 'min_child_weight': 0.13424014283416638, 'learning_rate': 0.14945689534042633, 'subsample': 0.8690041677979473, 'colsample_bylevel': 0.8428026817665201, 'colsample_bytree': 0.8817575629900145, 'reg_alpha': 0.02106728247448324, 'reg_lambda': 39.44246914091348, 'FLAML_sample_size': 393296}
Best r2 on validation data: 0.9992
Training duration of best run: 43.72 s
XGBClassifier(colsample_bylevel=0.8428026817665201,
              colsample_bytree=0.8817575629900145, grow_policy='lossguide',
              learning_rate=0.14945689534042633, max_depth=0, max_leaves=37,
              min_child_weight=0.13424014283416638, n_estimators=138, n_jobs=-1,
              reg_alpha=0.02106728247448324, reg_lambda=39.44246914091348,
              subsample=0.8690041677979473, tree_method='hist',
              use_label_encoder=False, verbosity=0)

And when I'm looking to retrain the model, I would like to use the same settings. I tried to define them using "custom_hp". But I'm having trouble in defining this. Is there any easy way to pick up these hyper-parameters and train on data?

**Comment by qingyun-wu** [comment, marked as answer]
Hi @vinay-k12,
Thank you for your question. There are two possible ways you can retain the model using the best config:
1. use[ `retrain_from_log`](https://microsoft.github.io/FLAML/docs/reference/automl/automl#retrain_from_log) if you logged the search trajectory. 
It allows you to retain a model using the best config in your log file. 
2. Set your best config to be the`starting_points` without specifying `max_iter` and `time_budget` . Find an example for using `starting_points` [here](https://github.com/microsoft/FLAML/blob/main/test/automl/test_warmstart.py#L148).  

Please let me know if you need any further help!

**Reply by qingyun-wu** [reply]
Hi @vinay-k12 ,

In approach option 2, you also need to specify the `estimator_list` to only include the learner you want to use, e.g., 'xgboost' in your example.  Please find an example showing how to use option 2 to retrain the model: https://github.com/qingyun-wu/FLAML/blob/qw-test/notebook/examples/retrain.ipynb

Thank you, and please let me know if you need further help!

**Reply by vinay-k12** [reply]
Awesome. That worked.

**Reply by sonichi** [reply]
> Hi @vinay-k12 ,
> 
> In approach option 2, you also need to specify the `estimator_list` to only include the learner you want to use, e.g., 'xgboost' in your example. Please find an example showing how to use option 2 to retrain the model: https://github.com/qingyun-wu/FLAML/blob/qw-test/notebook/examples/retrain.ipynb
> 
> Thank you, and please let me know if you need further help!

Could you please make the documentation more clear about retraining, model serialization etc.? Such questions are reoccurring.
